_posts/2020-06-27-lsp-explained.md

```diff
@@ -0,0 +1,183 @@
+
+```java
+class PaymentProcessor {
+  void process(OrderDetails orderDetails, PaymentInstrument paymentInstrument) {
+    try {
+      paymentInstrument.validate();
+      paymentInstrument.runFraudChecks();
+      paymentInstrument.sendToPaymentGateway();
+      saveToDatabase(orderDetails, paymentInstrument);
+    } catch (PaymentInstrumentInvalidException e) {
+      e.printStackTrace();
+    } catch (FraudDetectedException e) {
+      e.printStackTrace();
+    } catch (PaymentFailedException e) {
+      e.printStackTrace();
+    }
+  }
+
+  void saveToDatabase(OrderDetails orderDetails, PaymentInstrument paymentInstrument) {
+    String fingerprint = paymentInstrument.getFingerprint();
+    // save fingerprint and order details in DB
+  }
+}
+```
+
+Of course, in an actual production system, there would be many complex aspects to handle. The single processor class above might well be a bunch of classes in multiple packages across service and repository layers.
+
+All is well and our system is processing payments as expected. At some point as the business grows, the marketing team decides to introduce reward points to increase customer loyalty. Customers get a small number of reward points on each purchase. They can use the points to buy products on the website. Now there's a business need to support Rewards Card as a payment instrument in our payment system.
+
+Since Rewards Card is just another type of payment instrument, ideally we should be able to just add a ```RewardsCard``` class that extends ```PaymentInstrument``` and be done with it. But we find that adding it violates LSP! 
+
+There are no fraud checks against third-party systems for Rewards Cards. Details are not sent to payment gateways and there is no concept of fingerprint identifier for a Rewards Card. ```PaymentProcessor``` breaks as soon as we add ```RewardsCard```.
+
+To continue using our current class hierarchy, we might end up force-fitting this new class by overriding and providing empty, do-nothing implementations for ```runFraudChecks()``` and ```sendToPaymentGateway()```. 
+
+This would still break the application - we would get a ```NullPointerException``` from the repository layer ```saveToDatabase``` method since the fingerprint would be ```null```. We might be able to handle this just this one time as a special case in ```saveToDatabase``` by doing an ```instanceof``` check on the ```PaymentInstrument``` argument. But we know that it will never be this one time and it will never be this one place. 
+
+Soon our payment module codebase will be strewn with multiple checks and special cases to handle the problems created by the incorrect class model. We can imagine the pain this will cause each time we want to enhance the payment module. 
+
+For example, what if the business decides to accept bitcoins or some other cryptocurrency tomorrow? Or what if a new payment mode like Cash on Delivery is introduced?
+
```

Comment:
You're explaining an example of a violation of the LSP. Could you also add a section which explains how to do this better?

Reply:
Good catch, I have added a section with the fix now - I think this makes the article much better.